### PR TITLE
Remove redundant data member in favor of centralizing in DiffRequest. NFC

### PR DESCRIPTION
**Pull request by vgvassilev**
Partially addresses #721.

### CODE REVIEW

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 1]
```diff
@@ -258,12 +258,15 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
     assert(m_DiffReq == request);
 
-    // reverse mode plugins may have request mode other than
+    // FIXME: reverse mode plugins may have request mode other than
     // `DiffMode::reverse`, but they still need the `DiffMode::reverse` mode
     // specific behaviour, because they are "reverse" mode plugins.
-    m_Mode = DiffMode::reverse;
-    if (request.Mode == DiffMode::jacobian)
-      m_Mode = DiffMode::jacobian;
+    // assert(m_DiffReq.Mode == DiffMode::reverse ||
```
Comment (vaithak) [inline lib/Differentiator/ReverseModeVisitor.cpp:264]: Do we need these comments?

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 2]
```diff
@@ -258,12 +258,15 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
+    //        m_DiffReq.Mode == DiffMode::jacobian && "Unexpected Mode.");
+    if (m_DiffReq.Mode == DiffMode::error_estimation)
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+      const_cast<DiffRequest&>(m_DiffReq).Mode = DiffMode::reverse;
```
Comment (vaithak) [inline lib/Differentiator/ReverseModeVisitor.cpp:268]: I am not sure error estimation works, so can't judge if this is necessary.

File: lib/Differentiator/ReverseModeVisitor.cpp (hunk 1 above)
Comment (vgvassilev) [reply lib/Differentiator/ReverseModeVisitor.cpp:264]: Yes, that's a reminder to enable this once we address the fact that we should not change the mode of the DiffRequest.

File: lib/Differentiator/ReverseModeVisitor.cpp (hunk 2 above)
Comment (vgvassilev) [reply lib/Differentiator/ReverseModeVisitor.cpp:268]: Unfortunately without that line the error estimation test suite starts to fail.

**Pull request merged**